### PR TITLE
Add dotnet tool installer

### PR DESCRIPTION
**Pull request by sutser**

### CODE REVIEW

File: Plugins/BackbonePluginTemplatePlugin/BackbonePluginTemplatePluginGenerator.cs [hunk 1]
```diff
@@ -85,6 +88,7 @@ public void Generate()
 
         public void TearDown()
         {
+            _dotnetToolInstaller.InstallTool(SolutionDirectory, "backbone", null);
```
Comment (janhohenheim) [inline Plugins/BackbonePluginTemplatePlugin/BackbonePluginTemplatePluginGenerator.cs]: Doesn't belong to this PR

File: MetaGeneratorAbstractions/IDotnetToolInstaller.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Messerli.MetaGeneratorAbstractions
+{
+    public interface IDotnetToolInstaller
+    {
+        void InstallTool(string path, string toolName, string? version);
```
Comment (janhohenheim) [inline MetaGeneratorAbstractions/IDotnetToolInstaller.cs]: Add an overload without version and make this non-nullable in order to not pass awkward `null` values

File: MetaGenerator/DotnetToolInstaller.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                "tool",
+                "install",
+                toolName,
+            };
+            toolArguments.AddRange(GetVersionArguments(version));
+            return toolArguments;
+        }
+
+        private IEnumerable<string> GetVersionArguments(string? version)
+            => version is { }
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: if you add an identifier after `{}`, e.g. `ver`, you can use it instead of `version!` down below

File: MetaGenerator/DotnetToolInstaller.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        private IEnumerable<string> GetToolInstallArguments(string toolName, string? version)
+        {
+            var toolArguments = new List<string>
+            {
+                "tool",
+                "install",
+                toolName,
+            };
+            toolArguments.AddRange(GetVersionArguments(version));
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: you don't need to edit the list. Simply use [Enumerable.Concat](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.concat?view=netframework-4.8) to make this action immutable.

File: MetaGenerator/DotnetToolInstaller.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            _tools.RegisterTool(Dotnet, "dotnet.exe");
+            _tools.VerifyTools();
+        }
+
+        public void InstallTool(string path, string toolName, string? version)
+        {
+            const string dotnetToolsFileName = "dotnet-tools.json";
+            var dotnet = _tools.GetTool("dotnet.exe");
+
+            var hasDotnetToolManifest = Directory.GetFiles(path, dotnetToolsFileName, SearchOption.AllDirectories)
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: move this to own method

File: MetaGenerator/DotnetToolInstaller.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    public class DotnetToolInstaller : IDotnetToolInstaller
+    {
+        private const string Dotnet = "dotnet";
+
+        private readonly ITools _tools;
+
+        public DotnetToolInstaller(ITools tools)
+        {
+            _tools = tools;
+            _tools.RegisterTool(Dotnet, "dotnet.exe");
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: Move this to a constant

File: MetaGenerator/DotnetToolInstaller.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            _tools.VerifyTools();
+        }
+
+        public void InstallTool(string path, string toolName, string? version)
+        {
+            const string dotnetToolsFileName = "dotnet-tools.json";
+            var dotnet = _tools.GetTool("dotnet.exe");
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: Doesn't this expect the `name`, and not the executable? Does this call actually work?
See [ITools.cs](https://github.com/messerli-informatik-ag/meta-generator/blob/master/ToolLoaderAbstractions/ITools.cs)

File: MetaGenerator/DotnetToolInstaller.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+            var hasDotnetToolManifest = Directory.GetFiles(path, dotnetToolsFileName, SearchOption.AllDirectories)
+                .Any();
+
+            if (!hasDotnetToolManifest)
+            {
+                dotnet.Execute(new[] { "new", "tool-manifest" }, path);
+            }
+
+            dotnet.Execute(GetToolInstallArguments(toolName, version), path);
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: Move the execute calls into own methods. There's no benefit to sharing the `dotnet` object, as it will be used exactly once anyways.

**Review by janhohenheim — CHANGES_REQUESTED** [review]
As you don't need to actually mutate any collection of items, use arrays instead of lists

File: Plugins/BackbonePluginTemplatePlugin/BackbonePluginTemplatePluginGenerator.cs (hunk 1 above)
Comment (sutser) [reply Plugins/BackbonePluginTemplatePlugin/BackbonePluginTemplatePluginGenerator.cs]: Addressed in: [Remove installation of Backbone from this PR](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/2901ec651bce329cb3c3425f77b20e6f85c320a3)

File: MetaGeneratorAbstractions/IDotnetToolInstaller.cs (hunk 2 above)
Comment (sutser) [reply MetaGeneratorAbstractions/IDotnetToolInstaller.cs]: Addressed in: [Create overload method without version](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/408adf6e493e82f549351fb9fffd67e1639ecd60)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 7 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Change GetTools parameter](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/6c2c6c7af1eb17466e3800c578f57da8598c547a)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 6 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Add dotnet.exe constant](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/fc86e795204dfaa297217dac472fda9173d3e640)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 5 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Add HasDotNetToolManifest method](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/52822a9dd50cedaa42ad9206b8c61caefc4c980d)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 8 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Refactor DotnetToolInstaller](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/0bbe55b4481be7d90539bb1b81a03793c0b95cb9)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 4 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Refactor DotnetToolInstaller](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/0bbe55b4481be7d90539bb1b81a03793c0b95cb9)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 3 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Refactor DotnetToolInstaller](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/0bbe55b4481be7d90539bb1b81a03793c0b95cb9)

File: MetaGeneratorAbstractions/IDotnetToolInstaller.cs (hunk 2 above)
Comment (janhohenheim) [reply MetaGeneratorAbstractions/IDotnetToolInstaller.cs]: make the version non-nullable in the other overload as well.

File: MetaGenerator/DotnetToolInstaller.cs (hunk 4 above)
Comment (janhohenheim) [reply MetaGenerator/DotnetToolInstaller.cs]: You're modifying `arguments`. Don't mutate state if not absolutely necessary (functional programming). Return the `Concat` directly.

File: MetaGeneratorAbstractions/IDotnetToolInstaller.cs (hunk 2 above)
Comment (sutser) [reply MetaGeneratorAbstractions/IDotnetToolInstaller.cs]: Addressed in: [Refactor DotnetToolInstaller](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/0bbe55b4481be7d90539bb1b81a03793c0b95cb9)

File: MetaGenerator/DotnetToolInstaller.cs [hunk 9]
```diff
@@ -40,6 +37,14 @@ public void InstallTool(string path, string toolName, string? version)
             dotnet.Execute(GetToolInstallArguments(toolName, version), path);
         }
 
+        private static bool HasDotNetToolManifest(string path)
+        {
+            const string dotnetToolsFileName = "dotnet-tools.json";
+            var hasDotnetToolManifest = Directory.GetFiles(path, dotnetToolsFileName, SearchOption.AllDirectories)
+                .Any();
+            return hasDotnetToolManifest;
```
Comment (janhohenheim) [inline MetaGenerator/DotnetToolInstaller.cs]: return this directly

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: MetaGenerator/DotnetToolInstaller.cs (hunk 9 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Return HasDotnetToolManifest directly](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/ed06eda13e779b0a4bb1448479145e1830fb3741)

File: MetaGenerator/DotnetToolInstaller.cs (hunk 4 above)
Comment (sutser) [reply MetaGenerator/DotnetToolInstaller.cs]: Addressed in: [Return concated install arguments directly](https://github.com/messerli-informatik-ag/meta-generator/pull/24/commits/88572d872ea4092cd332c6496b41ee7f98df30e5)

**Review by janhohenheim — APPROVED** [review]
